src/CoreWCF.Primitives/tests/ServiceKnownTypeTests.cs

```diff
@@ -0,0 +1,223 @@
+            var channel = ServiceKnownTypeTests.CreateChannel(factory);
+            BaseMsg msg = channel.Ping("hello");
+            Assert.Equal("hello", msg.msg);
+            ServiceKnownTypeTests.CloseChannel(factory, channel);
+        }
+    }
+
+    internal class PingServiceEx : IPingServiceEx
+    {
+        public BaseMsg Ping(string msg)
+        {
+            return new DerivedMsg { msg = msg };
+        }
+
+    }
+
+    [ServiceContract]
+    [System.ServiceModel.ServiceContract]
+    public interface IPingServiceEx
+    {
+        [OperationContract]
+        [System.ServiceModel.OperationContract]
+        BaseMsg Ping(string msg);
+
+    }
+
+    internal class PingService1 : IPingService1
+    {
+        public BaseMsg Ping(string msg)
+        {
+            return new DerivedMsg { msg = msg };
+        }
+
+    }
+
+    [ServiceContract]
+    [System.ServiceModel.ServiceContract]
+    [ServiceKnownType(typeof(DerivedMsg))]
+    [System.ServiceModel.ServiceKnownType(typeof(DerivedMsg))]
+    public interface IPingService1
+    {
+        [OperationContract]
+        [System.ServiceModel.OperationContract]
+        BaseMsg Ping(string msg);
+
+    }
+
+    internal class PingService2 : IPingService2
+    {
+        public BaseMsg Ping(string msg)
+        {
+            return new DerivedMsg { msg = msg };
+        }
+
+    }
+
+    [ServiceContract]
+    [System.ServiceModel.ServiceContract]
+    [System.ServiceModel.ServiceKnownType(typeof(DerivedMsg))]
+    public interface IPingService2
+    {
+        [OperationContract]
+        [System.ServiceModel.OperationContract]
+        BaseMsg Ping(string msg);
+
+    }
+
+    internal class PingService3 : IPingService3
+    {
+        public BaseMsg Ping(string msg)
+        {
+            return new DerivedMsg { msg = msg };
+        }
+
+    }
+
+    [ServiceContract]
+    [System.ServiceModel.ServiceContract]
+    [ServiceKnownType("GetKnownTypes", typeof(DataProviderTypes))]
+    [System.ServiceModel.ServiceKnownType("GetKnownTypes", typeof(DataProviderTypes))]
+    public interface IPingService3
+    {
+        [OperationContract]
+        [System.ServiceModel.OperationContract]
+        BaseMsg Ping(string msg);
+
+    }
+
+    internal class PingService4 : IPingService4
+    {
+        public BaseMsg Ping(string msg)
+        {
+            return new DerivedMsg { msg = msg };
+        }
+
+    }
+
+    [ServiceContract]
+    [System.ServiceModel.ServiceContract]
+    [System.ServiceModel.ServiceKnownType("GetKnownTypes", typeof(DataProviderTypes))]
+    public interface IPingService4
+    {
+        [OperationContract]
+        [System.ServiceModel.OperationContract]
+        BaseMsg Ping(string msg);
+
+    }
+
+    [DataContract]
+    public class BaseMsg
+    {
+        [DataMember]
+        public string msg;
+    }
+
+    [DataContract]
+    public class DerivedMsg : BaseMsg
+    {
+    }
+
+    [DataContract]
+    public class DerivedMsg2 : BaseMsg
+    {
```

Comment:
DerivedMsg2 is never used

Reply:
Leftover from experimentation. Removed now.